### PR TITLE
Delete DL files on unsubscribe.

### LINKED ISSUE #16197: Add option on unsubscribe to retain DAT files
### What happened?

PR https://github.com/Chia-Network/chia-blockchain/pull/16182 deletes all the local DAT files on `unsubcribe` per https://github.com/Chia-Network/issue-tracker/issues/565

This is the new default behaviour

However, I think there should be an option to retain files that works like `unsubscribe` does today

### Version

2.1

### What platform are you using?

Windows

### What ui mode are you using?

GUI

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by fchirica**
Removes both locally stored full files and delta files when we unsubscribe to a delta layer store.

### CODE REVIEW

File: chia/data_layer/data_layer.py [hunk 1]
```diff
@@ -527,10 +532,22 @@ async def unsubscribe(self, tree_id: bytes32) -> None:
...
+                filenames.append(get_full_tree_filename(tree_id, root_hash, root.generation))
+                filenames.append(get_delta_filename(tree_id, root_hash, root.generation))
         async with self.subscription_lock:
             await self.data_store.unsubscribe(tree_id)
         await self.wallet_rpc.dl_stop_tracking(tree_id)
         self.log.info(f"Unsubscribed to {tree_id}")
+        for filename in filenames:
+            file_path = self.server_files_location.joinpath(filename)
+            if os.path.exists(file_path):
+                os.remove(file_path)
```
Comment (altendky) [inline chia/data_layer/data_layer.py:550]:
```suggestion
            try:
                file_path.unlink()
            except FileNotFoundError:
                pass
```

File: tests/core/data_layer/test_data_rpc.py [hunk 2]
```diff
@@ -2016,3 +2018,44 @@ async def test_issue_15955_deadlock(
...
+    )
+    async with init_data_layer(wallet_rpc_port=wallet_rpc_port, bt=bt, db_path=tmp_path) as data_layer:
+        data_rpc_api = DataLayerRpcApi(data_layer)
+        res = await data_rpc_api.create_data_store({})
+        root_hashes: List[bytes32] = []
+        assert res is not None
+        store_id = bytes32.from_hexstr(res["id"])
+        await farm_block_check_singelton(data_layer, full_node_api, ph, store_id)
+
+        for batch_count in range(10):
```
Comment (altendky) [inline tests/core/data_layer/test_data_rpc.py:2040]: I think a variable for the `10` would be helpful.  Then the `20` below could reference it for clarity.

```suggestion
        update_count = 10
        for batch_count in range(update_count):
```

File: tests/core/data_layer/test_data_rpc.py [hunk 3]
```diff
@@ -2016,3 +2018,44 @@ async def test_issue_15955_deadlock(
...
+            res = await data_rpc_api.batch_update({"id": store_id.hex(), "changelist": changelist})
+            update_tx_rec = res["tx_id"]
+            await farm_block_with_spend(full_node_api, ph, update_tx_rec, wallet_rpc_api)
+            await asyncio.sleep(10)
+            root_hash = await data_rpc_api.get_root({"id": store_id.hex()})
+            root_hashes.append(root_hash["hash"])
+
+        with os.scandir(data_layer.server_files_location) as entries:
+            filenames = {entry.name for entry in entries}
+            assert len(filenames) == 20
```
Comment (altendky) [inline tests/core/data_layer/test_data_rpc.py:2053]: May as well dedent as well.

```suggestion
            assert len(filenames) == 2 * update_count
```

File: tests/core/data_layer/test_data_rpc.py [hunk 4]
```diff
@@ -2016,3 +2018,44 @@ async def test_issue_15955_deadlock(
...
+            changelist = [{"action": "insert", "key": key.hex(), "value": value.hex()}]
+            res = await data_rpc_api.batch_update({"id": store_id.hex(), "changelist": changelist})
+            update_tx_rec = res["tx_id"]
+            await farm_block_with_spend(full_node_api, ph, update_tx_rec, wallet_rpc_api)
+            await asyncio.sleep(10)
+            root_hash = await data_rpc_api.get_root({"id": store_id.hex()})
+            root_hashes.append(root_hash["hash"])
+
+        with os.scandir(data_layer.server_files_location) as entries:
+            filenames = {entry.name for entry in entries}
```
Comment (altendky) [inline tests/core/data_layer/test_data_rpc.py:2052]: Feels like 
```suggestion
        filenames = {path.name for path in data_layer.server_files_location.iterdir()}
```

File: tests/core/data_layer/test_data_rpc.py [hunk 5]
```diff
@@ -2016,3 +2018,44 @@ async def test_issue_15955_deadlock(
...
+        await farm_block_check_singelton(data_layer, full_node_api, ph, store_id)
+
+        for batch_count in range(10):
+            key = batch_count.to_bytes(2, "big")
+            value = batch_count.to_bytes(2, "big")
+            changelist = [{"action": "insert", "key": key.hex(), "value": value.hex()}]
+            res = await data_rpc_api.batch_update({"id": store_id.hex(), "changelist": changelist})
+            update_tx_rec = res["tx_id"]
+            await farm_block_with_spend(full_node_api, ph, update_tx_rec, wallet_rpc_api)
+            await asyncio.sleep(10)
```
Comment (altendky) [inline tests/core/data_layer/test_data_rpc.py:2047]: This should probably be based on two times the read out configuration variable.  Or, preferably have a mechanism for waiting for the next periodic data management cycle.  But I'm ok leaving that for "later".  Would be nice though.

File: tests/core/data_layer/test_data_rpc.py [hunk 6]
```diff
@@ -2016,3 +2018,44 @@ async def test_issue_15955_deadlock(
...
+            filenames = {entry.name for entry in entries}
+            assert len(filenames) == 20
+            for generation, hash in enumerate(root_hashes):
+                assert get_delta_filename(store_id, hash, generation + 1) in filenames
+                assert get_full_tree_filename(store_id, hash, generation + 1) in filenames
+
+        res = await data_rpc_api.unsubscribe(request={"id": store_id.hex()})
+        with os.scandir(data_layer.server_files_location) as entries:
+            filenames = {entry.name for entry in entries}
+            assert len(filenames) == 0
```
Comment (altendky) [inline tests/core/data_layer/test_data_rpc.py:2061]: As above.

**Review by altendky — CHANGES_REQUESTED** [review]

**Review by altendky — DISMISSED** [review]
thanks

**Review by altendky — DISMISSED** [review]

**Review by emlowe — DISMISSED** [review]
Approved - well add separate issue to create a `retain` option

**Review by emlowe — APPROVED** [review]

**Comment by emlowe** [comment]
Coverage diff exception approved

**Pull request merged**